### PR TITLE
Improvements and bug fixes

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/arrays/implementations/map_array.ex [hunk 1]
```diff
@@ -4,31 +4,30 @@ defmodule Arrays.Implementations.MapArray do
...
     size = Keyword.get(options, :size, 0)
-    %MapArray{contents: construct(default, size), default: default}
+    %MapArray{contents: construct(default, size), default: default, size: size}
   end
 
   defp construct(_default, 0), do: %{}
 
   defp construct(default, size) do
-    0..size |> Enum.into(%{}, &{&1, default})
+    Enum.into(0..(size - 1), %{}, &{&1, default})
```
Comment (eksperimental) [inline lib/arrays/implementations/map_array.ex:18]: bug fix

File: lib/arrays/implementations/map_array.ex [hunk 2]
```diff
@@ -123,74 +122,79 @@ defmodule Arrays.Implementations.MapArray do
...
-    def extract(%MapArray{contents: contents}) when map_size(contents) == 0 do
+    def extract(%MapArray{size: 0}) do
       {:error, :empty}
     end
 
     def to_list(%MapArray{contents: contents}) do
-      for {_k, v} <- contents do
-        v
-      end
+      :maps.values(contents)
```
Comment (eksperimental) [inline lib/arrays/implementations/map_array.ex:197]: simplification

**Comment by Qqwy** [comment]
I don't think adding the size field to the struct should be necessary, as according to the documentation, `map_size`/[:maps.size/1](http://erlang.org/doc/man/maps.html#size-1) runs in constant time.

The bugfixes are however greatly appreciated! :slightly_smiling_face:

**Comment by eksperimental** [comment]
True that!
I will remove the size field then

**Comment by eksperimental** [comment]
@Qqwy please have a look again

**Comment by Qqwy** [comment]
Looks very nice! If you're happy, we could merge this.

**Comment by eksperimental** [comment]
> Looks very nice! If you're happy, we could merge this.

Please do so!

**Pull request merged**